include/session/attachments.hpp

```diff
@@ -0,0 +1,208 @@
+// crypto_secretstream encryption with a 32kiB chunk size.  We determine the padded size as follows,
+// given an input size N:
+//
+// - compute the total raw size M as N plus:
+// - 1 for the 'S' prefix (outside the encryption)
+// - 17 byte encryption overhead (crypto_secretstream_xchacha20poly1305_ABYTES = poly1305 MAC +
+//   1-byte tag) for every 32kiB (or piece thereof)
+// - 1 byte for the minimum padding size
+//
+// We then take keep the most-significant bit of M (i.e. reduce to the largest power of 2 <= M),
+// right-shift this by 5, and that round up to the next multiple of that padding factor.
+//
+// For example, for an input of 1MB (N=100000), we have an unpadded total size of 1000000+1+31*17 =
+// 1000528 (i.e. accounting for the 'S' identifier, and the 31 mac+tag values).  We then obtain the
+// highest power of two <= this (524288 = 2^19), right-shift by 5 to get 16384 (2^14), and then
+// round up the total size to the next multiple of that: 1015808.  Thus we require an additional
+// 15280 padding bytes, and so in total we get:
+//
+//   1 -- the leading (unencrypted) S
+// + 15279 × 0x00 -- leading padding bytes
+// + 1 × 0x01 -- final padding byte
+// + 1000000 -- encrypted file stream data (ignoring embedded mac+tags)
+// + 31 × 17 -- embedded mac+tags after every 32kiB of file stream data
+// = 1015808 final output.
+//
+// (Note that we always including at least one padding byte, and there are some complications in the
+// calculation as padding values get large enough to start inducing additional mac+tags; see the
+// implementation for details).
+size_t encrypted_padding(size_t data_size);
+
+/// API: crypto/attachment::encrypt
+///
+/// Encrypt an attachment for storage on the file server and distribution to other users using
+/// deterministic encryption where we use a cryptographically secure hash of the sending user's
+/// private key and file content to generate the encryption key/nonce pair.  The main advantage of
+/// this is deduplication: the same attachment uploaded by the same user will result in the same
+/// encrypted content, thus allowing deduplication of identical uploads on the file server.  This is
+/// particularly important for profile pictures, which are frequently re-uploaded to keep the
+/// attachment alive.
+///
+/// We currently always encrypt in chunks of (max) 32kiB via libsodium's crypto_secretstream API,
+/// and prefix the encrypted data with a 0x53 ('S') to indicate this.  Any other value of the first
+/// byte is reserved for possible alternative future encryption mechanisms.
+///
+/// We prepend padding of at least 1 byte before the actual data, by prepending (PADDING-1) 0x00
+/// bytes followed by a single 0x01 byte to the actual data stream; this data is discard when
+/// decrypting.
+///
+/// Inputs:
+/// - `seed` -- the 32-byte seed of the sender; it is recommended that this be the 32-byte Session
```

Comment:
Just to confirm - by this you mean the seed used to generate the users `ed25519` key pair?